File: py/translation_aspose_routing_test.py
"""
Standalone regression test for the NEW Translation-service Aspose
routing wired into py/server.py:
  /api/translation/analyze-strategy -> _handle_translation_analyze_strategy
  /api/translation/process-aspose   -> _handle_translation_process_aspose

Per explicit direction: table/background documents route to Aspose
(structure-aware conversion + real in-place LLM translation, via
aspose_test_pipeline.run_full_test - already tested through the admin
Aspose test route); everything else uses the existing client-side
pdf.js text-layer pipeline (js/engine-translation.js), unchanged.

This deliberately reuses ocr_router.analyze_pdf_for_ocr_strategy
directly (not a second copy of the same table/background-detection
logic) and aspose_test_pipeline.run_full_test directly (not a second
Aspose+translate pipeline) - this test confirms BOTH the reuse is wired
correctly AND the underlying decision is still correct for Translation's
own real documents.

Run: python3 translation_aspose_routing_test.py [path/to/table_pdf.pdf]
The second PDF (with real tables/background) is optional - if not
given, only the plain-text routing check runs (still meaningful: it's
exactly the shape of document Translation processes every day).
"""
import os
import re
import sys

sys.path.insert(0, os.path.dirname(__file__))
import ocr_router  # noqa: E402


def assert_(cond, label):
    if not cond:
        print("FAIL:", label)
        sys.exit(1)
    print("PASS:", label)


def run(plain_pdf, table_pdf=None):
    print("=== Test 1: plain-text PDF (Translation's everyday case) -> lightweight (pdf.js) ===")
    analysis = ocr_router.analyze_pdf_for_ocr_strategy(plain_pdf)
    assert_(analysis["strategy"] == "lightweight", "Plain document routes to lightweight, not Aspose - " + analysis["reason"][:80])

    print("\n=== Test 2: server.py wiring - both new routes registered ===")
    server_src = open(os.path.join(os.path.dirname(__file__), "server.py")).read()
    assert_('"/api/translation/analyze-strategy": self._handle_translation_analyze_strategy' in server_src,
            "/api/translation/analyze-strategy is registered in the URL routing table")
    assert_('"/api/translation/process-aspose": self._handle_translation_process_aspose' in server_src,
            "/api/translation/process-aspose is registered in the URL routing table")

    print("\n=== Test 3: handler bodies reuse the EXISTING, already-tested functions (not reimplementations) ===")
    def extract_method(name):
        start = server_src.find("def " + name + "(")
        assert start != -1, name + " not found"
        # crude but sufficient: grab up to the next top-level "    def "
        next_def = server_src.find("\n    def ", start + 1)
        return server_src[start:next_def if next_def != -1 else start + 4000]

    analyze_src = extract_method("_handle_translation_analyze_strategy")
    assert_("ocr_router.analyze_pdf_for_ocr_strategy(" in analyze_src,
            "_handle_translation_analyze_strategy reuses ocr_router.analyze_pdf_for_ocr_strategy directly (no duplicate table/background detection logic)")

    aspose_src = extract_method("_handle_translation_process_aspose")
    assert_("asp_test.run_full_test(" in aspose_src,
            "_handle_translation_process_aspose reuses aspose_test_pipeline.run_full_test directly (the same, already-tested Aspose+translate pipeline the admin test route uses)")
    assert_("AsposeNotConfiguredError" in aspose_src,
            "_handle_translation_process_aspose handles the not-configured case explicitly rather than letting it surface as a raw crash")

    print("\n=== Test 4: neither new route requires Admin/Developer role (real users must be able to call these) ===")
    assert_("_require_role" not in analyze_src, "_handle_translation_analyze_strategy has no role restriction")
    assert_("_require_role" not in aspose_src, "_handle_translation_process_aspose has no role restriction")

    if table_pdf:
        print("\n=== Test 5: real table/background PDF -> aspose ===")
        analysis2 = ocr_router.analyze_pdf_for_ocr_strategy(table_pdf)
        assert_(analysis2["strategy"] == "aspose", "Table/background document correctly routes to aspose - " + analysis2["reason"][:80])

    print("\nALL TESTS PASSED")


if __name__ == "__main__":
    default_plain = "/home/claude/work/ocr_debug/s36.pdf"
    plain = sys.argv[1] if len(sys.argv) > 1 else default_plain
    table = sys.argv[2] if len(sys.argv) > 2 else None
    if not os.path.exists(plain):
        print("Plain-text test PDF not found:", plain)
        sys.exit(1)
    run(plain, table)

File: js/translation_aspose_routing_test.js
// Test for the Translation service's NEW automatic Aspose-vs-pdf.js
// routing (js/app.js), replacing the old manual "With OCR" checkbox
// entirely, per explicit direction: a source PDF with tables or
// background color goes through Aspose (structure-aware conversion +
// real in-place LLM translation); everything else uses the existing
// client-side pdf.js text-layer pipeline, falling back automatically
// to vision-based OCR for genuinely scanned PDFs - mirroring
// ocr-service.js's own fallback pattern exactly, rather than a manual
// toggle deciding vision-vs-text-layer.

const fs = require('fs');
const path = require('path');

const src = fs.readFileSync(path.join(__dirname, 'app.js'), 'utf8');

function assert(cond, label) { if (!cond) { console.log('FAIL:', label); process.exitCode = 1; } else console.log('PASS:', label); }

// The old manual checkbox and its supporting variables are genuinely gone
// from the reachable code, not just renamed.
assert(!src.includes("getElementById('translationHybridCheck')"), 'No remaining reference to the removed translationHybridCheck checkbox element');
assert(!src.includes('const translationHybridMode ='), 'The old translationHybridMode constant declaration is removed');

// The new routing calls both real endpoints.
assert(src.includes("fetch('/api/translation/analyze-strategy'"), 'Calls the new /api/translation/analyze-strategy endpoint');
assert(src.includes("fetch('/api/translation/process-aspose'"), 'Calls the new /api/translation/process-aspose endpoint');

// The three-way strategy outcome drives naming/extension, not a binary hybridMode.
assert(src.includes("actualStrategyUsed === 'aspose'"), 'Output naming distinguishes the aspose outcome');
assert(src.includes("actualStrategyUsed === 'vision_ocr_fallback'"), 'Output naming distinguishes the vision-OCR-fallback outcome');
assert(src.includes("actualStrategyUsed === 'client_text_layer'") || src.includes("actualStrategyUsed = 'client_text_layer'"), 'The plain client-side text-layer outcome is tracked as its own named state');

// The automatic scanned-PDF fallback mirrors ocr-service.js's own pattern
// (catch the specific "scanned/image-based" signal, fall back to
// buildHybridDocxBlob) rather than a manual toggle.
const translationFlowStart = src.indexOf("if (translationStrategy === 'aspose')");
assert(translationFlowStart !== -1, 'The new translationStrategy dispatch exists');
const translationFlowSrc = src.slice(translationFlowStart, translationFlowStart + 7000);
assert(translationFlowSrc.includes('looksLikeScanSignal'), 'The scanned-PDF signal is checked, same as ocr-service.js');
assert(translationFlowSrc.includes('buildHybridDocxBlob'), 'Falls back to buildHybridDocxBlob automatically on that signal');
assert(translationFlowSrc.includes('buildOfflineDocxBlob'), 'Tries buildOfflineDocxBlob (pdf.js text-layer) first, as the default path');

// Billing: the Aspose path (no per-page event stream) must still charge
// correctly, using the page count already known from analyze-strategy.
assert(translationFlowSrc.includes('translationStrategyAnalysis') && translationFlowSrc.includes('page_count'),
  'Aspose path bills using the real page count from the analyze-strategy call (not left at zero for lack of per-page events)');

// A failed analyze-strategy call must not block translation entirely -
// same defensive default as OCR's own analyzeOcrStrategy().
const analyzeCallIdx = src.indexOf("fetch('/api/translation/analyze-strategy'");
const analyzeSurrounding = src.slice(Math.max(0, analyzeCallIdx - 800), analyzeCallIdx + 200);
assert(analyzeSurrounding.includes("let translationStrategy = 'lightweight'"), 'translationStrategy defaults to lightweight before the analysis call, so a failed request still lets translation proceed');

console.log(process.exitCode ? '\nSOME TESTS FAILED' : '\nALL TESTS PASSED');
